readme.md
## JS Loading Overlay
[![Build Status](https://img.shields.io/github/workflow/status/muhdfaiz/js-loading-overlay/JS%20Loading%20overlay)](https://github.com/muhdfaiz/js-loading-overlay/actions?query=branch%3Amaster)
[![Coverage Status](https://coveralls.io/repos/github/muhdfaiz/js-loading-overlay/badge.svg?branch=master)](https://coveralls.io/github/muhdfaiz/js-loading-overlay?branch=master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/muhdfaiz/js-loading-overlay/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/muhdfaiz/js-loading-overlay/?branch=master) 
![Downloads](https://img.shields.io/npm/dt/js-loading-overlay)
[![Version](https://img.shields.io/npm/v/js-loading-overlay)](https://www.npmjs.com/package/js-loading-overlay)
![Size](https://img.shields.io/bundlephobia/min/js-loading-overlay)

Display loading overlay easily and beautifully in your application. JS loading overlay using awesome spinner from [Load Awesome](https://github.danielcardoso.net/load-awesome/animations.html). It contains 52 spinner you can choose.

## Features

- **Full page loading overlay.** Able to show full page loading overlay. 

- **Loading overlay inside container.** Able to display the loading overlay inside the container

- **Lock scroll.** Able to lock the scroll during loading overlay active.

## Documentation

For detail of installation instructions, in-depth usage, please take a look at the [documentation](https://js-loading-overlay.muhdfaiz.com/).

### How To Install
```markdown
NPM
npm install --save js-loading-overlay

Yarn
yarn add js-loading-overlay

Use without NPM or Yarn
import 'js-loading-overlay'
JsLoadingOverlay.show({'spinnerIcon': 'triangle-skew-spin'});
```

### How To Use
```markdown
import 'js-loading-overlay'
JsLoadingOverlay.show({'spinnerIcon': 'triangle-skew-spin'});

or

<script src="js-loading-overlay.min.js"></script>
JsLoadingOverlay.show({'spinnerIcon': 'triangle-skew-spin'});
```

### Show loading overlay with default options

```markdown
JsLoadingOverlay.show();
```

### Show loading overlay with custom options

```markdown
JsLoadingOverlay.show({
  'overlayBackgroundColor': '#666666',
  'overlayOpacity': 0.6,
  'spinnerIcon': 'ball-atom',
  'spinnerColor': '#000',
  'spinnerSize': '2x',
  'overlayIDName': 'overlay',
  'spinnerIDName': 'spinner',
  'spinnerZIndex':99999,
  'overlayZIndex':99998
});
```

### Hide Loading Overlay

```markdown
JsLoadingOverlay.hide();
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to Semantic Versioning.

## 1.2.0 - 2020-12-05

- Add option to specify custom overlay z-index.
- Add option to specify custom spinner z-index.

## 1.1.0 - 2020-12-02

- Add option to adjust the spinner position for X axis and Y axis.
- Add option to display loading overlay in the container instead of full page only.
- Add option to lock screen when loading overlay has been shown.

## 1.0.5 - 2020-10-13

- Fixed coverall issue.

## 1.0.4 - 2020-10-08

- Fixed error when calling hide method during loading overlay not ective.

## 1.0.3 - 2020-04-03

- Fixed type main filename in package.json. Cause the user who install using NPM and YARN having an error when using it.

## 1.0.2 - 2020-01-08

- Fixed overlay and spinner position.

## 1.0.1 - 2020-11-12

- Use travis CI to run test and check coverage.
- Update demo page and readme.

## 1.0.0 - 2019-11-10

- Initial release! 🎉

test/loading-overlay.test.js
const loadingOverlay = require('../src/js-loading-overlay');
const XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;

beforeEach(() => {
    document.body.innerHTML = '';
    document.head.innerHTML = '';
    loadingOverlay.options = {
        'overlayBackgroundColor': '#666666',
        'overlayOpacity': 0.6,
        'spinnerIcon': 'ball-circus',
        'spinnerColor': '#000',
        'spinnerSize': '3x',
        'overlayIDName': 'overlay',
        'spinnerIDName': 'spinner',
        'offsetY': 0,
        'offsetX': 0,
        'lockScroll': false,
        'containerID': null,
        'overlayZIndex': 99998,
        'spinnerZIndex': 99999,
    };
});

describe('generateEmptyDivElement function', () => {
   test('It should generate 6 empty div element', () => {
       let numberOfEmptyDivToGenerate = 6;

        let emptyDivElements = loadingOverlay.generateEmptyDivElement(numberOfEmptyDivToGenerate);

        document.body.innerHTML = emptyDivElements;
        expect(document.getElementsByTagName('div').length).toEqual(numberOfEmptyDivToGenerate);
    });

    test('It should generate 0 empty div element', () => {
        let numberOfEmptyDivToGenerate = 0;

        let emptyDivElements = loadingOverlay.generateEmptyDivElement(numberOfEmptyDivToGenerate);

        expect(emptyDivElements).toEqual('');
    });
});

describe('setSpinnerStylesheetURL function', () => {
    test('It should set correct stylesheet for spinner icon ball-spin-clockwise', () => {
        let spinnerIcon = 'ball-spin-clockwise';

        loadingOverlay.options = {
            'spinnerIcon': spinnerIcon,
        };

        loadingOverlay.setSpinnerStylesheetURL();

        expect(loadingOverlay.stylesheetBaseURL + spinnerIcon + '.min.css').toEqual(loadingOverlay.spinnerStylesheetURL);
    });

    test('It should set valid stylesheet url for spinner icon ball-atom', async () => {
        let spinnerIcon = 'ball-atom';

        loadingOverlay.options = {
            'spinnerIcon': spinnerIcon,
        };

        loadingOverlay.setSpinnerStylesheetURL();

        expect(loadingOverlay.spinnerStylesheetURL).toEqual(loadingOverlay.stylesheetBaseURL + spinnerIcon + '.min.css');

        async function urlExists(url){
            var http = new XMLHttpRequest();
            http.open('HEAD', url, false);
            http.send();
            return http.status == 200;
        }

        let stylesheetURLExist = await urlExists(loadingOverlay.spinnerStylesheetURL);
        expect(stylesheetURLExist).toBeTruthy();
    });

    test('It should set incorrect stylesheet when spinner icon is invalid', async () => {
        let spinnerIcon = 'test-icon';

        loadingOverlay.options = {
            'spinnerIcon': spinnerIcon,
        };

        loadingOverlay.setSpinnerStylesheetURL();

        expect(loadingOverlay.spinnerStylesheetURL).toEqual(loadingOverlay.stylesheetBaseURL + spinnerIcon + '.min.css');

        async function urlExists(url){
            var http = new XMLHttpRequest();
            http.open('HEAD', url, false);
            http.send();
            return http.status == 200;
        }

        let stylesheetURLExist = await urlExists(loadingOverlay.spinnerStylesheetURL);
        expect(stylesheetURLExist).toBeFalsy();
    });
});

describe('addSpinnerStylesheet function', () => {
    test('It should add spinner stylesheet for spinner icon ball-spin-clockwise in the header', () => {
        let spinnerIcon = 'ball-spin-clockwise';

        loadingOverlay.options = {
            'spinnerIcon': spinnerIcon,
        };

        loadingOverlay.addSpinnerStylesheet();

        expect(document.getElementsByTagName('link')[0].getAttribute('href')).toBe(loadingOverlay.spinnerStylesheetURL);
    });
});

describe('generateSpinnerElement function', () => {
    test('It should generate correct spinner element', () => {
        let spinnerIcon = 'ball-spin-clockwise';
        let spinnerColor = 'rgb(236, 236, 236)';
        let spinnerSize = 'small';

        loadingOverlay.options = {
            'spinnerIcon': spinnerIcon,
            'spinnerColor': spinnerColor,
            'spinnerSize': spinnerSize,
        };

        loadingOverlay.generateSpinnerElement();

        document.body.innerHTML = loadingOverlay.spinner;
        expect(document.getElementsByClassName('la-' + spinnerIcon).length).toEqual(1);
        expect(document.getElementsByClassName('la-' + spinnerSize).length).toEqual(1);
        expect(document.getElementsByClassName('la-' + spinnerSize)[0].style.color).toEqual(spinnerColor);
    })
});

describe('generateOverlayElement function', () => {
    test('It should generate correct overlay element in full page', () => {
        let overlayBackgroundColor = 'rgb(0, 0, 0)';
        let overlayOpacity = "0.6";
        let spinnerIcon = 'ball-spin-clockwise';
        let spinnerColor = 'rgb(236, 236, 236)';
        let spinnerSize = 'small';
        let overlayIDName = 'overlay';
        let spinnerIDName = 'spinner';

        loadingOverlay.options = {
            'overlayBackgroundColor': overlayBackgroundColor,
            'overlayOpacity': overlayOpacity,
            'spinnerIcon': spinnerIcon,
            'spinnerColor': spinnerColor,
            'spinnerSize': spinnerSize,
            'overlayIDName': overlayIDName,
            'spinnerIDName': spinnerIDName,
        };

        loadingOverlay.show();

        // Expect spinner element.
        expect(document.getElementsByClassName('la-' + spinnerSize).length).toEqual(1);
        expect(document.getElementsByClassName('la-' + spinnerSize)[0].style.color).toEqual(spinnerColor);

        // Expect overlay element.
        expect(document.getElementById(overlayIDName)).not.toBeNull();
        expect(document.getElementById(overlayIDName).style.backgroundColor).toEqual(overlayBackgroundColor);
        expect(document.getElementById(overlayIDName).style.opacity).toBe(overlayOpacity);
    })

    test('It should generate correct overlay element in container', () => {
        let overlayBackgroundColor = 'rgb(0, 0, 0)';
        let overlayOpacity = "0.6";
        let spinnerIcon = 'ball-spin-clockwise';
        let spinnerColor = 'rgb(236, 236, 236)';
        let spinnerSize = 'small';
        let overlayIDName = 'overlay';
        let spinnerIDName = 'spinner';

        loadingOverlay.options = {
            'overlayBackgroundColor': overlayBackgroundColor,
            'overlayOpacity': overlayOpacity,
            'spinnerIcon': spinnerIcon,
            'spinnerColor': spinnerColor,
            'spinnerSize': spinnerSize,
            'overlayIDName': overlayIDName,
            'spinnerIDName': spinnerIDName,
            'containerID': 'table'
        };

        document.body.innerHTML = '<div id="table"><table><tr><th>Company</th><th>Contact</th><th>Country</th></tr><tr><td>Test Company</td><td>Muhammad Faiz</td><td>Malaysia</td></tr></table></div>'
        loadingOverlay.show();

        // Expect overlay and spinner element inside the container.
        expect(document.getElementById(loadingOverlay.options.containerID).contains(document.getElementById(loadingOverlay.options.overlayIDName))).toBe(true);
        expect(document.getElementById(loadingOverlay.options.containerID).contains(document.getElementById(loadingOverlay.options.spinnerIDName))).toBe(true);

        // Expect spinner element.
        expect(document.getElementsByClassName('la-' + spinnerIcon).length).toEqual(1);
        expect(document.getElementsByClassName('la-' + spinnerSize).length).toEqual(1);
        expect(document.getElementsByClassName('la-' + spinnerSize)[0].style.color).toEqual(spinnerColor);

        // Expect overlay element.
        expect(document.getElementById(overlayIDName)).not.toBeNull();
        expect(document.getElementById(overlayIDName).style.backgroundColor).toEqual(overlayBackgroundColor);
        expect(document.getElementById(overlayIDName).style.opacity).toBe(overlayOpacity);
    })
});

describe('setOptions function', () => {
    test('It should override default options when options parameter exist.', () => {
        let options = {
            'overlayBackgroundColor': '#000',
            'overlayOpacity': 1,
            'spinnerIcon': 'ball-beat',
            'spinnerColor': '#fff',
            'spinnerSize': '2x',
            'overlayIDName': 'overlay-id',
            'spinnerIDName': 'spinner-id',
            'offsetX': 10,
            'offsetY': 20,
            'lockScroll': true,
            'containerID': 'table',
            'overlayZIndex': 99998,
            'spinnerZIndex': 99999,
        };

        loadingOverlay.setOptions(options);

        expect(loadingOverlay.options.overlayBackgroundColor).toEqual(options.overlayBackgroundColor);
        expect(loadingOverlay.options.overlayOpacity).toEqual(options.overlayOpacity);
        expect(loadingOverlay.options.spinnerIcon).toEqual(options.spinnerIcon);
        expect(loadingOverlay.options.spinnerColor).toEqual(options.spinnerColor);
        expect(loadingOverlay.options.spinnerSize).toEqual(options.spinnerSize);
        expect(loadingOverlay.options.overlayIDName).toEqual(options.overlayIDName);
        expect(loadingOverlay.options.spinnerIDName).toEqual(options.spinnerIDName);
        expect(loadingOverlay.options.offsetX).toEqual(options.offsetX);
        expect(loadingOverlay.options.offsetY).toEqual(options.offsetY);
        expect(loadingOverlay.options.lockScroll).toEqual(options.lockScroll);
        expect(loadingOverlay.options.containerID).toEqual(options.containerID);
        expect(loadingOverlay.options.spinnerZIndex).toEqual(options.spinnerZIndex);
        expect(loadingOverlay.options.overlayZIndex).toEqual(options.overlayZIndex);
    });

    test('It should not override default options when options parameter not exist.', () => {
        loadingOverlay.setOptions();

        expect(loadingOverlay.options.overlayBackgroundColor).toEqual('#666666');
        expect(loadingOverlay.options.overlayOpacity).toEqual(0.6);
        expect(loadingOverlay.options.spinnerIcon).toEqual('ball-circus');
        expect(loadingOverlay.options.spinnerColor).toEqual('#000');
        expect(loadingOverlay.options.spinnerSize).toEqual('3x');
        expect(loadingOverlay.options.overlayIDName).toEqual('overlay');
        expect(loadingOverlay.options.spinnerIDName).toEqual('spinner');
        expect(loadingOverlay.options.offsetX).toEqual(0);
        expect(loadingOverlay.options.offsetY).toEqual(0);
        expect(loadingOverlay.options.lockScroll).toEqual(false);
        expect(loadingOverlay.options.containerID).toEqual(null);
    });
});

describe('hide function', () => {
    test('It should hide the loading overlay.', () => {
        let defaultOptions = {
            'overlayBackgroundColor': '#666666',
            'overlayOpacity': 0.6,
            'spinnerIcon': 'ball-circus',
            'spinnerColor': '#000',
            'spinnerSize': '3x',
            'overlayIDName': 'overlay',
            'spinnerIDName': 'spinner'
        };

        loadingOverlay.show(defaultOptions);

        expect(document.getElementById(loadingOverlay.options.overlayIDName)).toBeTruthy();
        expect(document.getElementById(loadingOverlay.options.spinnerIDName)).toBeTruthy();
        expect(document.getElementsByTagName('link')[0].getAttribute('href')).toBe(loadingOverlay.spinnerStylesheetURL);

        loadingOverlay.hide();
        expect(document.getElementById(loadingOverlay.options.overlayIDName)).toBeNull();
        expect(document.getElementById(loadingOverlay.options.spinnerIDName)).toBeNull();
        expect(document.getElementsByTagName('link').length).toEqual(0);
    });

    test('It should not return an error when the loading overlay not shown.', () => {
        loadingOverlay.hide();
    });

    test('It should unlock back the scroll after hide.', () => {
        document.body.style.position = 'relative';
        document.body.style.width = '500px';
        document.body.style.top = '100px';

        let defaultOptions = {
            'overlayBackgroundColor': '#666666',
            'overlayOpacity': 0.6,
            'spinnerIcon': 'ball-circus',
            'spinnerColor': '#000',
            'spinnerSize': '3x',
            'overlayIDName': 'overlay',
            'spinnerIDName': 'spinner',
            'lockScroll': true,
        };

        loadingOverlay.show(defaultOptions);
        expect(document.body.style.overflow).toEqual('hidden');
        expect(document.documentElement.style.overflow).toEqual('hidden');

        loadingOverlay.hide();
        expect(document.body.style.overflow).toEqual('');
        expect(document.documentElement.style.overflow).toEqual('');
    });
});

describe('show function', () => {
    test('It should show the loading overlay in full page.', () => {
        let defaultOptions = {
            'overlayBackgroundColor': '#666666',
            'overlayOpacity': 0.6,
            'spinnerIcon': 'ball-circus',
            'spinnerColor': '#000',
            'spinnerSize': '3x',
            'overlayIDName': 'overlay',
            'spinnerIDName': 'spinner'
        };

        loadingOverlay.show(defaultOptions);

        expect(document.querySelectorAll('body > #' + defaultOptions.overlayIDName)).toBeTruthy();
        expect(document.getElementById(loadingOverlay.options.overlayIDName)).toBeTruthy();
        expect(document.getElementById(loadingOverlay.options.spinnerIDName)).toBeTruthy();
        expect(document.getElementsByTagName('link')[0].getAttribute('href')).toBe(loadingOverlay.spinnerStylesheetURL);
    });

    test('It should show the loading overlay in container page.', () => {
        document.body.innerHTML = '<div id="table"><table><tr><th>Company</th><th>Contact</th><th>Country</th></tr><tr><td>Test Company</td><td>Muhammad Faiz</td><td>Malaysia</td></tr></table></div>'

        let defaultOptions = {
            'overlayBackgroundColor': '#666666',
            'overlayOpacity': 0.6,
            'spinnerIcon': 'ball-circus',
            'spinnerColor': '#000',
            'spinnerSize': '3x',
            'overlayIDName': 'overlay',
            'spinnerIDName': 'spinner',
            'containerID': 'table'
        };

        loadingOverlay.show(defaultOptions);

        expect(document.querySelectorAll('body > #' + defaultOptions.overlayIDName).length === 0).toBe(true);

        // Expect overlay and spinner element inside the container.
        expect(document.getElementById(loadingOverlay.options.containerID).contains(document.getElementById(loadingOverlay.options.overlayIDName))).toBe(true);
        expect(document.getElementById(loadingOverlay.options.containerID).contains(document.getElementById(loadingOverlay.options.spinnerIDName))).toBe(true);

        expect(document.getElementById(loadingOverlay.options.overlayIDName)).toBeTruthy();
        expect(document.getElementById(loadingOverlay.options.spinnerIDName)).toBeTruthy();
        expect(document.getElementsByTagName('link')[0].getAttribute('href')).toBe(loadingOverlay.spinnerStylesheetURL);
    });

    test('It should show the loading overlay in container page.', () => {
        document.body.innerHTML = '<div id="table"><table><tr><th>Company</th><th>Contact</th><th>Country</th></tr><tr><td>Test Company</td><td>Muhammad Faiz</td><td>Malaysia</td></tr></table></div>'

        let defaultOptions = {
            'overlayBackgroundColor': '#666666',
            'overlayOpacity': 0.6,
            'spinnerIcon': 'ball-circus',
            'spinnerColor': '#000',
            'spinnerSize': '3x',
            'overlayIDName': 'overlay',
            'spinnerIDName': 'spinner',
            'containerID': 'table'
        };

        loadingOverlay.show(defaultOptions);

        expect(document.querySelectorAll('body > #' + defaultOptions.overlayIDName).length === 0).toBe(true);

        // Expect overlay and spinner element inside the container.
        expect(document.getElementById(loadingOverlay.options.containerID).contains(document.getElementById(loadingOverlay.options.overlayIDName))).toBe(true);
        expect(document.getElementById(loadingOverlay.options.containerID).contains(document.getElementById(loadingOverlay.options.spinnerIDName))).toBe(true);

        expect(document.getElementById(loadingOverlay.options.overlayIDName)).toBeTruthy();
        expect(document.getElementById(loadingOverlay.options.spinnerIDName)).toBeTruthy();
        expect(document.getElementsByTagName('link')[0].getAttribute('href')).toBe(loadingOverlay.spinnerStylesheetURL);
    });

    test('It should show the loading overlay with custom offset of spinner position.', () => {
        document.body.innerHTML = ''

        let defaultOptions = {
            'offsetY': '-10px',
            'offsetX': '20px',
        };

        loadingOverlay.show(defaultOptions);

        expect(document.querySelectorAll('body > #' + defaultOptions.overlayIDName).length === 0).toBe(true);
        expect(document.body.innerHTML).toContain('calc(50% + ' + defaultOptions.offsetY);
        expect(document.body.innerHTML).toContain('calc(50% + ' + defaultOptions.offsetX);
    });

    test('It should show the loading overlay with custom z-index.', () => {
        document.body.innerHTML = ''

        let defaultOptions = {
            'spinnerZIndex': 50,
            'overlayZIndex': 49
        };

        loadingOverlay.show(defaultOptions);

        expect(document.querySelectorAll('body > #' + defaultOptions.overlayIDName).length === 0).toBe(true);

        expect(document.getElementById(loadingOverlay.options.overlayIDName).style.zIndex).toBe(defaultOptions.overlayZIndex.toString());
        expect(document.getElementById(loadingOverlay.options.spinnerIDName).style.zIndex).toBe(defaultOptions.spinnerZIndex.toString());
    });
});


webpack.config.js
const path = require("path");
const TerserPlugin = require('terser-webpack-plugin');
const env = process.env.NODE_ENV;

module.exports = (env, options) => {
    return {
        devtool: "source-map",
        target: "async-node",
        entry: ["./src/js-loading-overlay.js"],
        output: {
            filename: options.mode == 'development' ? "dist/js-loading-overlay.js" : "dist/js-loading-overlay.min.js",
            path: path.resolve(__dirname, ''),
        },
        module: {
            rules: [{
                test: /\.js$/,
                exclude: /node_modules/,
                loader: 'babel-loader',
            }]
        },
        optimization: {
            minimizer: [
                new TerserPlugin({
                    extractComments: true,
                    cache: true,
                    parallel: true,
                    sourceMap: true, // Must be set to true if using source-maps in production
                    terserOptions: {
                        // https://github.com/webpack-contrib/terser-webpack-plugin#terseroptions
                        extractComments: 'all',
                        compress: {
                            drop_console: true,
                        },
                        output: {
                            comments: false,
                        },
                    }
                }),
            ],
        }
    }
};

src/js-loading-overlay.js
class JSLoadingOverlay
{
    constructor() {
        this.options = {
            'overlayBackgroundColor': '#666666',
            'overlayOpacity': 0.6,
            'spinnerIcon': 'ball-circus',
            'spinnerColor': '#000',
            'spinnerSize': '3x',
            'overlayIDName': 'overlay',
            'spinnerIDName': 'spinner',
            'offsetY': 0,
            'offsetX': 0,
            'lockScroll': false,
            'containerID': null,
            'spinnerZIndex': 99999,
            'overlayZIndex': 99998
        };
        this.stylesheetBaseURL = 'https://cdn.jsdelivr.net/npm/load-awesome@1.1.0/css/';
        this.spinner = null;
        this.spinnerStylesheetURL = null;
        this.numberOfEmptyDivForSpinner = {
            'ball-8bits' : 16,
            'ball-atom': 4,
            'ball-beat': 3,
            'ball-circus': 5,
            'ball-climbing-dot': 1,
            'ball-clip-rotate': 1,
            'ball-clip-rotate-multiple': 2,
            'ball-clip-rotate-pulse': 2,
            'ball-elastic-dots': 5,
            'ball-fall': 3,
            'ball-fussion': 4,
            'ball-grid-beat': 9,
            'ball-grid-pulse': 9,
            'ball-newton-cradle': 4,
            'ball-pulse': 3,
            'ball-pulse-rise': 5,
            'ball-pulse-sync': 3,
            'ball-rotate': 1,
            'ball-running-dots': 5,
            'ball-scale': 1,
            'ball-scale-multiple': 3,
            'ball-scale-pulse': 2,
            'ball-scale-ripple': 1,
            'ball-scale-ripple-multiple': 3,
            'ball-spin': 8,
            'ball-spin-clockwise' : 8,
            'ball-spin-clockwise-fade' : 8,
            'ball-spin-clockwise-fade-rotating' : 8,
            'ball-spin-fade' : 8,
            'ball-spin-fade-rotating' : 8,
            'ball-spin-rotate' : 2,
            'ball-square-clockwise-spin' : 8,
            'ball-square-spin' : 8,
            'ball-triangle-path' : 3,
            'ball-zig-zag' : 2,
            'ball-zig-zag-deflect' : 2,
            'cog' : 1,
            'cube-transition' : 2,
            'fire' : 3,
            'line-scale' : 5,
            'line-scale-party' : 5,
            'line-scale-pulse-out' : 5,
            'line-scale-pulse-out-rapid' : 5,
            'line-spin-clockwise-fade' : 8,
            'line-spin-clockwise-fade-rotating' : 8,
            'line-spin-fade' : 8,
            'line-spin-fade-rotating' : 8,
            'pacman' : 6,
            'square-jelly-box' : 2,
            'square-loader' : 1,
            'square-spin' : 1,
            'timer' : 1,
            'triangle-skew-spin' : 1,
        }
    }

    /**
     * Show loading overlay including the spinner.
     *
     * @param options
     */
    show(options) {
        // Override default options if options parameter exist.
        this.setOptions(options);

        // Add spinner stylesheet from https://cdn.jsdelivr.net in head.
        this.addSpinnerStylesheet();

        // Generate spinner html element.
        this.generateSpinnerElement();

        if (this.options.lockScroll) {
            document.body.style.overflow = 'hidden';
            document.documentElement.style.overflow = 'hidden';
        }

        // Generate overlay html element in full page.
        this.generateAndAddOverlayElement();
    }

    /**
     * Hide Loading Overlay
     */
    hide() {
        // Unlock scroll.
        if (this.options.lockScroll) {
            document.body.style.overflow = '';
            document.documentElement.style.overflow = '';
        }

        var stylesheet = document.getElementById('loading-overlay-stylesheet');

        // Remove stylesheet.
        if (stylesheet) {
            stylesheet.disabled = true;
            stylesheet.parentNode.removeChild(stylesheet);

            document.getElementById(this.options.overlayIDName).remove();
            document.getElementById(this.options.spinnerIDName).remove();
        }
    }

    /**
     * Override default options with user specified options.
     *
     * @param options
     */
    setOptions(options) {
        if (typeof options !== 'undefined') {
            for (let key in options) {
                this.options[key] = options[key];
            }
        }
    }

    /**
     * Generate overlay html element in full page.
     * - class name
     * - background color
     * - opacity
     *
     * @returns {string}
     */
    generateAndAddOverlayElement() {
        let left = '50%';

        // Check if spinner X offset not zero
        if (this.options.offsetX !== 0) {
            left = 'calc(50% + ' + this.options.offsetX + ')'
        }

        let top = '50%';

        // Check if spinner Y offset not zero
        if (this.options.offsetY !== 0) {
            top = 'calc(50% + ' + this.options.offsetY + ')'
        }

        // Generate overlay html element in container.
        if (this.options.containerID && document.body.contains(document.getElementById(this.options.containerID))) {
            let loadingOverlay = `<div id="${this.options.overlayIDName}" style="display: block !important; position: absolute; top: 0; left: 0; overflow: auto; opacity: ${this.options.overlayOpacity}; background: ${this.options.overlayBackgroundColor}; z-index: 50; width: 100%; height: 100%;"></div><div id="${this.options.spinnerIDName}" style="display: block !important; position: absolute; top: ${top}; left: ${left}; -webkit-transform: translate(-50%); -ms-transform: translate(-50%); transform: translate(-50%); z-index: 9999;">${this.spinner}</div>`

            let containerID = document.getElementById(this.options.containerID);

            containerID.style.position = 'relative';
            containerID.insertAdjacentHTML('beforeend', loadingOverlay);
            return;
        }

        let loadingOverlay = `<div id="${this.options.overlayIDName}" style="display: block !important; position: fixed; top: 0; left: 0; overflow: auto; opacity: ${this.options.overlayOpacity}; background: ${this.options.overlayBackgroundColor}; z-index: ${this.options.overlayZIndex}; width: 100%; height: 100%;"></div><div id="${this.options.spinnerIDName}" style="display: block !important; position: fixed; top: ${top}; left: ${left}; -webkit-transform: translate(-50%); -ms-transform: translate(-50%); transform: translate(-50%); z-index: ${this.options.spinnerZIndex};">${this.spinner}</div>`

        // Insert the overlay html element in body.
        document.body.insertAdjacentHTML("beforeend", loadingOverlay);
    }

    /**
     * Generate spinner html element. Spinner html element must follow template from https://github.danielcardoso.net/load-awesome/
     */
    generateSpinnerElement()
    {
        let emptyDivKey = Object.keys(this.numberOfEmptyDivForSpinner).find(element => element === this.options.spinnerIcon);

        let emptyDivElement = this.generateEmptyDivElement(this.numberOfEmptyDivForSpinner[emptyDivKey]);

        this.spinner = `<div style="color: ${this.options.spinnerColor}" class="la-${ this.options.spinnerIcon } la-${this.options.spinnerSize}">${ emptyDivElement }</div>`
    }

    /**
     * Add spinner stylesheet tag in head section.
     * All stylesheet URL are from CDN. Using Load Awesome CSS Spinner.
     *
     * @see https://github.danielcardoso.net/load-awesome/
     * @see https://www.jsdelivr.com/package/npm/load-awesome?path=css
     */
    addSpinnerStylesheet() {
        this.setSpinnerStylesheetURL();

        let link = document.createElement('link');
        link.setAttribute('id', 'loading-overlay-stylesheet');
        link.setAttribute('rel', 'stylesheet');
        link.setAttribute('type', 'text/css');
        link.setAttribute('href', this.spinnerStylesheetURL);

        document.getElementsByTagName('head')[0].appendChild(link);
    }

    /**
     * Set spinner stylesheet based on the icon specify in options.
     * All stylesheet URL are from CDN. Using Load Awesome CSS Spinner.
     *
     * @see https://github.danielcardoso.net/load-awesome/
     * @see https://www.jsdelivr.com/package/npm/load-awesome?path=css
     *
     * @returns {string}
     */
    setSpinnerStylesheetURL() {
        this.spinnerStylesheetURL = this.stylesheetBaseURL + this.options.spinnerIcon + '.min.css'
    }

    /**
     * Generate empty div element for spinner element.
     *
     * @param numberOfEmptyDiv
     * @returns {string}
     */
    generateEmptyDivElement(numberOfEmptyDiv) {
        let emptyDivElement = '';

        for (let i=1; i <= numberOfEmptyDiv; i++) {
            emptyDivElement += '<div></div>'
        }

        return emptyDivElement;
    }
}

window.JsLoadingOverlay = new JSLoadingOverlay();

module.exports = JsLoadingOverlay;